mocha-shared-behaviors/mocha-shared-behaviors.md

```diff
@@ -0,0 +1,471 @@
+    expect(user.name.first).to.equal('tobi');
+  });
+  // other tests
+}
+
+/// user.test.js
+describe('User', () => {
+  let user;
+
+  beforeEach(() => {
+    user = new User('tobi', 'holowaychuk');
+  });
+
+  shouldBehaveLikeAUser(user);
+});
+```
+
+Let's run this and...:boom: it fails!
+
+```
+TypeError: Cannot read property 'name' of undefined
+  at Context.name (test/helpers.js:5:17)
+```
+
+This is because Mocha identifies and "records" your test suite first, and _then_ runs your callbacks. So here, it runs
+`beforeEach` and `shouldBehaveLikeAUser` (`user` being undefined at this point) and only _then_ `beforeEach.fn` and `it.fn`.
+
+## "All-in-one"
+
+One solution is to move the `beforeEach` in `shouldBehaveLikeAUser`.
+
+```javascript
+/// helpers.js
+export function shouldBehaveLikeAUser(buildUserFn, { firstName, lastName, fullname }) {
+  let userLike;
+
+  beforeEach(() => {
+    userLike = buildUserFn();
+  });
+  
+  it('should have .name.first', () => {
+    expect(userLike.name.first).to.equal(firstName);
+  });
+  // other tests
+};
+
+/// user.test.js
+describe('User', () => {
+  shouldBehaveLikeAUser(() => new User("tobi", "holowaychuk"), {
+    firstName: "tobi",
+    lastName: "holowaychuk",
+    fullname: 'tobi holowaychuk'
+  });
+});
+
+/// admin.test.js
+describe('Admin', () => {
+  shouldBehaveLikeAUser(() => new Admin("tobi", "holowaychuk"), {
+    firstName: "tobi",
+    lastName: "holowaychuk",
+    fullname: 'tobi holowaychuk'
+  });
+});
+```
+
+Here, nothing is "hidden." Just by looking at the signature, we understand that `shouldBehaveLikeAUser` will test that the constructor you gave will fit the "User" behavior definition. This can be enhanced by adding a JSDoc @param or some TypeScript.
+
+And it's self-sufficient. No side effects or closure requirements here.
+
+**But**, it has its own logic. For example, if you add the first argument to the `Admin` constructor, it fails, because you introduced a bug in the shared behavior, and there is a chance you won't see it directly because you didn't define a spec for that.
+
+And it's completely isolated! You can't reuse `userLike`, and have to repeat yourself.
+
+```javascript
+it('should be an .admin', () => {
+  expect(new Admin().admin).to.be.true;
+});
+```
+
+**Conclusion**
+
+You should use this solution if, and only if, you absolutely need specific setup or tear down which are completely isolated from the other behaviors, and if the required parameters aren't too complex.
+
+## "one-by-one"
+
+One way to solve the issues the previous method raise is by defining the shared behaviors one by one, spec by spec, like this:
+
+```javascript
+export const expectUserLike = user => ({
+  toHaveNameFirstAs: mock => {
+    expect(user.name.first).to.equal(mock);
+  },
+  toHaveNameLastAs: mock => {
+    expect(user.name.last).to.equal(mock);
+  },
+  toHaveFullnameThatReturnAs: (...mock) => {
+    expect(user.fullname()).to.equal(mock.join(' '));
```

Comment:
imho this is not nice... e.g. you expect that the implementation logic is the same as the test logic...
imho it's always better to handwrite what you expect

so I would sugguest
```
expect(user.fullname()).to.equal(mock);
```